PG-1294 WAL: encrypt segment ranges (#64)
percona/postgres: contrib/pg_tde/documentation/docs/setup.md

## WAL encryption configuration (tech preview)

Before turning WAL encryption on, you must first create a principal key for WAL.

Here's what to do:

> 1. Create pg_tde extesion if it is not exists:

Reviewer: ```suggestion
1. Create the `pg_tde` extension if it doesn't exist yet:
```

Reviewer: Do we need to mention where to create the extension? It'll be created for the current db, right? Will it affect WAL encryption anyhow?

Author: We need to create extension to use functions from the next steps (pg_tde_add_global_key_provider_vault_v2 etc). If we assume that extension should've been created on some previous steps than we can commit this step. But it is safer have it I think

    ```sql
    CREATE EXTENSION IF NOT EXISTS pg_tde;
    ```

> 2. Set up the key provider for WAL encryption

Reviewer 2: a key provider

        This setup is intended for development and stores the keys unencrypted in the specified data file.    

        ```sql

>         SELECT pg_tde_add_global_key_provider_file('provider-name','/path/to/the/keyring/data.file');

Reviewer 2: default_key_provide also works

Author: Should we elaborate it here? Or just create a reference here to possible options once all those functions are documented?

        ```

> 3. Create principal key

Reviewer 2: for WAL

> 3. Create principal key for WAL

Reviewer: ```suggestion
3. Create a principal key for WAL
```

    
    ```sql
    SELECT pg_tde_set_server_principal_key('principal-key', 'provider-name');
    ```

> 4. Enable WAL level encryption using the `ALTER SYSTEM` command. You need the privileges of the superuser to run this command:

Reviewer 2: you need superuser privileges

    ```sql
    ALTER SYSTEM SET pg_tde.wal_encrypt = on;
    ```

5. Restart the server to apply the changes.

    * On Debian and Ubuntu:    

       ```sh

>        sudo systemctl restart postgresql.service

Reviewer 2: `restart postgresql` is enough

       ```
    
    * On RHEL and derivatives

       ```sh
       sudo systemctl restart postgresql-17
       ```

> Now WAL files start to be encrypted for both encrypted and unencrypted tables.

Reviewer 2: for everything, not just tables.

Reviewer: I meant "encrypted and unencrypted tables with user data". Meaning if a user has dbs with pg_tde created against them and those without pg_tde, WAL will be encrypted for all of them

Reviewer 2: WAL is not specific to tables, any database operation possibly can be logged there. For example, when we create an internal key, we write it into the WAL. Now it will be encrypted, and it's not a table. Everything that is written into the WAL will be encrypted.

Author: ```
Now all data in WAL files start to be encrypted, including data for unencrypted table.
```
?
